### PR TITLE
Markdownlint: custom rules documentation

### LINKED ISSUE #27498: Markdownlint: TOP001 rule is unclear
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [X] Would you like to work on this issue?

### Describe your suggestion

The TOP001 rule regarding descriptive text labels is confusing. The detail of the rule, `Links must have a more descriptive text label` does not indicate what is wrong with the text, making it difficult to fix the issue without looking at the rule implementation. Additionally, the rule (and the other custom rules) do not have documentation. 

I would do the following:
- Change the detail from `Links must have a more descriptive text label.`  to `Link text should not contain 'this' or 'here'`, 
- Add a documentation file using the same [structure as the default rules](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md). It will contain a description of the rule, what it is for, examples of text that break the rule, and corrected versions of the examples. 
- Link the rule's documentation to the rule through the `information` prop. This enables the Odinite to open the documentation's link when the rule is triggered if he's using the VSC plugin.

### Path

Other / NA

### Lesson Url

https://github.com/TheOdinProject/curriculum/blob/main/markdownlint/TOP001_descriptiveLinkText/TOP001_descriptiveLinkText.js

### (Optional) Discord Name

Cake

### (Optional) Additional Comments

```[tasklist]
### Acceptance criteria
- [ ] https://github.com/TheOdinProject/curriculum/issues/27546
- [ ] https://github.com/TheOdinProject/curriculum/issues/27547
```

### LINKED ISSUE #27547: Markdownlint: add documentation to each custom rule

### PR DESCRIPTION
**Pull request by cakegod**
## Because

Our markdownlint custom rules can be clarified with documentation.

## This PR

- Adds documentation to markdownlint custom rules.

## Issue

Closes #27547

## Additional Information

Do we put all documentation files in a separate `docs` directory or do we put each documentation file in the corresponding rule directory?

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by nikitarevenco** [comment]
I think a great addition would be to also include relevant documentation for the actual markdown lint errors that are being thrown on lesson preview. For example, if there's a descriptive-link error, it would display the message from the docs

**Comment by cakegod** [comment]
@nikitarevenco That shouldn't be necessary if the rules provide a descriptive error message that explains the issue and how to resolve it. If you believe that the existing error messages could be improved, please feel free to open an issue.

**Review by thatblindgeye — COMMENTED** [review]
Thanks for doing this! 🔥 🔥 🔥 

I'm gonna give this a more proper review within the week, but just wanted to mention a couple quick thoughts.

- It'd be great if we could update our workflow to omit these MD files from the linter. If you'd feel up to that I'll say go ahead, but if you're not comfortable with that would you mind just opening a followup issue?
- What do you think about a README in this new directory, just to briefly explain the purpose of these docs and maybe to link to the default rules provided by markdownlint?

**Comment by cakegod** [comment]
I'll do that. 

Do I open a new PR to update the workflow?

**Comment by thatblindgeye** [comment]
Just the docs directory. There are MD files in each individual rule/test directory that we want to have checked as our "testing".

And you can update the workflow in this PR if you feel comfortable, otherwise you can open a separate PR if you'd rather.

File: markdownlint/docs/TOP001.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# `TOP001` - Descriptive Link Text
+
+Tags: `accessibility`, `links`
+
+Aliases: `descriptive-link-text`
+
+This rule is triggered when a link has a text label that uses the words "this" or "here", for example:
+
+```markdown
+You can read about it [here](https://example.com).
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP001.md:10]: Just to make this one match the fixed version a little more:

```suggestion
You can read more about variables in JavaScript [here](https://example.com).
```

File: markdownlint/docs/TOP001.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+[This article](https://example.com) provides more information about closures.
+```
+
+To fix this issue, remove the words "here" and "this" and use more descriptive text that clearly conveys the purpose or content of the link:
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP001.md:14]:
```suggestion
To fix this issue, remove the words "here" and "this" and use more descriptive text that clearly conveys the purpose or content of the link. You may need to also update the surrounding text. The following code block demonstrates how the links from above could be fixed:
```

File: markdownlint/docs/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# TOP Custom Markdownlint Rules
+
+This directory contains documentation for our custom rules for linting Markdown files using the markdownlint tool. These rules supplement the [default rules provided by markdownlint](https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md) to enforce our [curriculum's layout style](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md).
```
Comment (thatblindgeye) [inline markdownlint/docs/README.md:3]:
```suggestion
This directory contains documentation for our custom rules for linting Markdown files using the markdownlint tool. These rules supplement the [default rules provided by markdownlint](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md) to enforce our [curriculum's layout style](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md).
```

File: markdownlint/docs/TOP002.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+To fix this issue, remove the inline code from the heading and rephrase it to convey the same meaning without using code:
+
+```markdown
+## Using console.log for Debugging
+```
+
+## Rationale
+
+Headings should be concise and descriptive, providing a clear overview of the section's content. Including inline code in headings can make them less readable and may not effectively convey the purpose of the section to users.
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP002.md:21]: @KevinMulhern do you recall the reason we started omitting inline code in headings? I thought it was due to how the ID's are created on the heading elements not playing well with backticks, but can't actually recall.

File: markdownlint/docs/TOP003.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+Aliases: `default-section-content`
+
+This rule is triggered when a section does not have the expected default content or structure. The affected sections include:
+
+- Lesson Overview
+- Assignment
+- Knowledge Check
+- Additional Resources
+
+Below are some examples of errors and their solutions:
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP003.md:14]: I don't think we really need this line.

File: markdownlint/docs/TOP003.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+## Contains Content Before or After List
+
+There should be no content before or after the unordered list, other than the default content.
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP003.md:18]:
```suggestion
There should be no content before or after the unordered list of items in the "Lesson overview", "Knowledge check", and "Additional resources" sections, other than the default content.
```

File: markdownlint/docs/TOP003.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+```markdown
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- LO item
+```
+
+Incorrect (missing the list after the `Additional Resources` heading):
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP003.md:78]: Let's combine this and the below incorrect LO example, we don't really need to show both.

File: markdownlint/docs/TOP003.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+```markdown
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- An item
+- Another item
+```
+
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP003.md:135]: Let's add a heading for the Assignment section not having the expected div wrapper.

**Review by thatblindgeye — CHANGES_REQUESTED** [review]
In addition to the below comments, let's update the headings in each Md to use sentence case rather than capitalizing each word.

File: markdownlint/docs/TOP003.md (hunk 8 above)
Comment (cakegod) [reply markdownlint/docs/TOP003.md:135]: It seems we're missing a test for that.

File: markdownlint/docs/TOP002.md (hunk 4 above)
Comment (thatblindgeye) [reply markdownlint/docs/TOP002.md:21]: @cakegod so it seems like the reason we may have stopped using inline code in headings was (at least partially) due to how the ID and anchor elements were rendered. The backticks/inline code would cause things to render a little wonky (duplication of the heading title).

I think what you have here is fine for now, though, as it helps prevent someone from wanting to make a heading that's verbose code (rather than something that would be fine, such as wrapping specific keywords in backticks e.g. "The `lang` attribute").

Comment (cakegod) [reply markdownlint/docs/TOP002.md:21]: Should we add that info to the rationale?

Comment (thatblindgeye) [reply markdownlint/docs/TOP002.md:21]: We can do that in a followup. It may not be an issue any longer, but requires a little more investigating that I wouldn't want holding this PR up for.

File: markdownlint/docs/TOP003.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+## Missing "Assignment" section div wrapper
+
+The "Assignment" section ordered list should be wrapped in a div with the correct attributes.
```
Comment (thatblindgeye) [inline markdownlint/docs/TOP003.md:118]:
```suggestion
The "Assignment" section should contain a div with the correct attributes. Note that not all assignment content needs to be wrapped in this div, but it must at least exist in this section.
```

**Review by thatblindgeye — CHANGES_REQUESTED** [review]

File: markdownlint/docs/TOP003.md (hunk 8 above)
Comment (thatblindgeye) [reply markdownlint/docs/TOP003.md:140]: Can you open a followup issue for this and ping me in it?

Comment (cakegod) [reply markdownlint/docs/TOP003.md:140]: I've included the test in this [commit](https://github.com/TheOdinProject/curriculum/pull/27545/commits/ee3302aa214d192a05e0f74b755e2bd93174d780). Do we still open a new issue?

File: markdownlint/docs/TOP003.md (hunk 9 above)
Comment (cakegod) [reply markdownlint/docs/TOP003.md:118]: Not all assignment content?

File: markdownlint/docs/TOP003.md (hunk 8 above)
Comment (thatblindgeye) [reply markdownlint/docs/TOP003.md:140]: Oop nevermind 😆

File: markdownlint/docs/TOP003.md (hunk 9 above)
Comment (thatblindgeye) [reply markdownlint/docs/TOP003.md:118]: Not currently no. Maybe something to revisit, but right now we're just checking whether at least one of these div wrappers exists in the Assignment section.

**Review by thatblindgeye — APPROVED** [review]
